WL update misbehaving

cstamas opened the pull request:
This pull contains fix for multiple WL related issues: two related to WL performance, one for polluting logs and one about maintaing WL for hosted reposes.

First, group WL's were double updated in some cases (like group member addition or removal). This resulted in correct results but the "wl update" operation was done twice, redundantly.

Second, on Nexus boot groups were "inited" multiple times, redundantly. The "init loop" of WL was traversing hosted and proxy reposes only, leaving group init to happen as "side effect" of cascade operation (when a repo WL changes, it cascades to all groups where repo in question is member). But this also meant, that one group WL was updated (published, event fired, but no WL content change after 1st pass) as many times as many members it had. Now, the boot process is modified, to not use "cascade" and explicitly init group repositories too.

Third, logs were polluted on boot with M1 reposes not supported by WL feature (if you had non shadow M1 reposes in your config).

Fourth, when a parent folder was deleted (parent of a WL entry for hosted repo), the WL was not properly maintained.

Related issues:
https://issues.sonatype.org/browse/NEXUS-5602
https://issues.sonatype.org/browse/NEXUS-5608
https://issues.sonatype.org/browse/NEXUS-5618
https://issues.sonatype.org/browse/NEXUS-5623

CI:
https://bamboo.zion.sonatype.com/browse/NXO-OSSF39-1

cstamas committed:
NEXUS-5608: Group WL was updated more then once in some cases.

EventDispatcher initially tried to be more selective when to invoke
update on given group, but due to recent changes it ended up with
two almost identical methods (if repo in question was group).

This change simply "unhooks" one of those two.

cstamas committed:
NEXUS-5602: WL init add significant startup time.

This change now splits "WL init at boot" from "WL init
at repo addition on runtime" cases. Before, both were
doing same, but due to "cascade", one Nexus with lot of
repo and/or groups booted slower than before.

Now the "cascade" is not done in case of "boot" but on
"repo add at runtime" case is still done (and is needed).
On boot, extra step is done to explicitly init group reposes
WL, instead relying on having them updated implicitly due to
cascade.

This also meant, while former solution was correct, it was
sloppy as same group was updated over and over again on boot
process.

cstamas committed:
NEXUS-5602: WL init causes long startup time

Added UT, just to realize that initial fix is wrong, due to
concurrency issues of direct updates of group WLs.

Second attempt should be more involved but clearer.

Added UT covering both cases: boot-init and runtime
repo addition, verifying that WL updates does not happen more than
needed.

cstamas committed:
Merge remote-tracking branch 'origin/master' into nexus-5608-group-wl-double-update

cstamas committed:
NEXUS-5608: Making IT more reliable

Hosted wl update happens async just like proxies, and it was
just by chance that this IT was passing (but also failing sometimes)

Now with wait for async WL jobs it should behave better.

jdillon commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/maven/wl/internal/EventDispatcher.java:
```diff
@@ -339,8 +339,8 @@ public void onRepositoryGroupMembersChangedEvent( final RepositoryGroupMembersCh
-    @Subscribe
-    @AllowConcurrentEvents
+    // @Subscribe
+    // @AllowConcurrentEvents
```
Please add some additional comments as to why these are commented out. If this is not temporary, then these should be removed not commented.

jdillon commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/maven/wl/internal/WLManagerImpl.java:
```diff
@@ -225,10 +254,41 @@ public void shutdown()
     {
-        doInitializeWhitelist( mavenRepository );
+        getLogger().debug( "Initializing white-list of newly added {}",
+            RepositoryStringUtils.getHumanizedNameString( mavenRepository ) );
```
Curious, why isn't the default toString() of these objects rendering "humanized" forms?  This "humanization" of the str repr would be better done there and would allow for slf4j {} optimization to be used to avoid string concat for logs like this when the level is not enabled.  Otherwise stuff like this should be guarded.

I recommend you consider changing the toString() to render what is appropriate for logging, in this case whatever the {{getHumanizedNameString()}} is doing and avoid copying this code all over the place.

cstamas replied:
I believe you are right. This is DEBUG log, and I already added toString for reposes (eases debugging but is handy also in DEBUG logs). Will change,

cstamas committed:
NEXUS-5608: Removing unneeded methods.

These methods were added very early at WL imple
stage, but during development they proved unneeded
as the one "onConfigUpdated" handles all we need.

cstamas committed:
Logging cleanup.

Getting rid of extensive use of RSU to format
logged repo name and relying on toString method
of repo implementations in DEBUG and lower TRACE
logging.

cstamas committed:
Added missing punctuation :P

cstamas commented:
new CI
https://bamboo.zion.sonatype.com/browse/NXO-OSSF39-5

cstamas committed:
NEXUS-5602: Fixing the test

The code is correct but test was incorrect as it asserted
against one possible outcome of updates (due to concurrency).

The point is that -- depending how bg jobs were
executed -- group was updated one or twice. One thing is
sure, group has to be updated AT LEAST ONCE.

Adjusted assertions accordingly, to not make this test
fail in case when group is updated only once, as that is
also valid outcome.

cstamas commented:
CI +1d (after tightening a correct but unstable UT)
https://bamboo.zion.sonatype.com/browse/NXO-OSSF39-6

This pull is ready to be merged.

cstamas committed:
NEXUS-5618: M1 reposes are polluting logs on boot

The init step of WLManager did not change for Maven2
content class, only for the MavenRepo "facet". This
is not enough, as on boot M1 reposes will
be tried to init wrongfully. That is not
causing any problem except polluting logs
with stack traces.

We had something similar already fixed, but it was
about repositories put out of service...

cstamas committed:
NEXUS-5623: Hosted WL not updated on parent removal

WL for hosted repositories is not properly maintained
when a parent directory is removed (parent of WL entry).

For example, when your WL has entry like "/org/sonatype"
and you delete folder "/org".

Moreover, the UT testing this was broken too, as it was
testing this but actually missed the point: comments
stated one thing but assertions were asserting the
opposite

kellyrob99 commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/maven/wl/internal/WLManagerImpl.java:
```diff
@@ -225,13 +256,42 @@ public void shutdown()
+            // and publish
+            updateWhitelist( mavenRepository );
+            getLogger().info( "Initializing non-existing white-list of newly added {}",
+                RepositoryStringUtils.getHumanizedNameString( mavenRepository ) );
```
We don't need the RepositoryStringUtils call anymore do we?

cstamas replied:
Yes, we had a discussion about this with @jdillon ... the Repository.toString is implemented since 2.3 and is fine to be used in logging.

Here, the RSU is used only -- as level is INFO and is among 1st things user sees when Nexus boots -- to have it "aligned" with other INFO messages, where repositories are read up from config, as those lines are using RSU also.

I don't have a strong stance about use or non-use of RSU in logging, but according to previous comments from Jason, all DEBUG and TRACE level logging (at least in WL) has removed RSU use to not have the "protected" (if logger.isDebugEnabled...), but in case of these INFO level logs I don't consider them as problem.

kellyrob99 replied:
Thanks for the explanation

kellyrob99 commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/maven/wl/internal/WLManagerImpl.java:
```diff
@@ -225,13 +256,42 @@ public void shutdown()
+        catch ( Exception e )
+        {
+            getLogger().warn( "Problem during white-list initialisation of newly added {}",
+                RepositoryStringUtils.getHumanizedNameString( mavenRepository ), e );
```
RepositoryStringUtils again

kellyrob99 commented on nexus-core/src/test/java/org/sonatype/nexus/proxy/maven/wl/internal/WLMaintenanceTest.java:
```diff
@@ -224,11 +224,11 @@ public void smoke()
         {
             final List<String> entries = getEntriesOf( mavenRepository );
-            assertThat( entries.size(), equalTo( 5 ) );
+            assertThat( entries.size(), equalTo( 4 ) );
```
can use hasSize(4) instead for better messaging in case of failure

kellyrob99 commented:
+1

cstamas committed:
Applied Kelly's comments.

Using hasSize in assertions for better messaging
in case of UT failure.

cstamas commented:
@kellyrob99 Thanks for comments, hasSize change applied.

cstamas commented:
CI +1'd
https://bamboo.zion.sonatype.com/browse/NXO-OSSF39-7

cstamas committed:
UT testing WL content on member change.

There was no UT directly testing for correct WL
content on group member changes (only "update propagation"
was tested).

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/maven/wl/internal/WLManagerImpl.java:
```diff
@@ -169,26 +169,57 @@ public WLManagerImpl( final EventBus eventBus, final ApplicationStatusSource app
+                    doInitializeWhitelistOnStartup( mavenRepository );
                 }
             }
         }
```
Why do we need to iterate over hosted/proxy and group repositories separately here? doInitializeWhitelistOnStartup only loads existing prefix file and does not depend on group repository members as far as I can tell.

ifedorenko commented on nexus-core/src/test/java/org/sonatype/nexus/proxy/maven/wl/internal/WLUpdatePropagationGroupUpdatesTest.java:
```diff
@@ -0,0 +1,258 @@
+ * 
+ * @author cstamas
+ */
+public class WLUpdatePropagationGroupUpdatesTest
```
Can you add tests for group-of-groups?

jdillon commented:
VERY STRANGE src/test/java/core/whitelist/WhitelistWithHostedRepositoryIT.java

cstamas committed:
Extended WL update tests to GoG

Group of group checks introduced in the UTs.

cstamas commented:
CI:
https://bamboo.zion.sonatype.com/browse/NXO-OSSF39-9

cstamas commented:
CI +1'd
https://bamboo.zion.sonatype.com/browse/NXO-OSSF39-9

cstamas commented:
Discussed over the phone with @ifedorenko, where he blessed this pull. Merging it.
